pkg/gather/clusterconfig/clusterconfig.go

```diff
@@ -57,6 +62,8 @@ var (
 
 	// logTailLines sets maximum number of lines to fetch from pod logs
 	logTailLines = int64(100)
+
+	imageHostRegex = regexp.MustCompile(`(^|\.)(openshift\.org|registry\.redhat\.io|registry\.access\.redhat\.com)$`)
```

Comment:
This is really questionable.  Why are you doing this, vs fetching the SHAs?  You cannot look at image URLs and assume anything about them, since customers may have both intercepting proxies, be using a registry with pull through capabilities, or use the internal registry.

Instead, you should be returning the actual running SHA of the image (from status) and it should be the responsibility of the team to post process this data.

I'm not happy that this change went in without wider review, but code in OpenShift shouldn't be making these assumptions (that pull spec is meaningful).

Reply:
@smarterclayton The point of this enhancement was by no means to catch all cases of all running images from all registries. To the contrary, the goal was to collect as little data as possible while gathering some information about containers that are most likely running an image published by Red Hat.

The reason why we're collecting containers from `Spec` instead of `Status` is because we were going through the JSON alphabetically and the containers seemed to be the same. Of course, this is an easy fix. The regular containers seemed the most important, which is why they're the only ones being collected. Similarly, this is an easy fix and we'll add init and ephemeral containers in the fix.

We spent a month repeatedly asking various OpenShift engineers, as well as people from Product Security, who asked for this enhancement, what's the best way to filter Red Hat images. No one was able to tell us much. We were told to look for the image labels, but those don't really seem to be easily accessible via the K8s API. Filtering based on URL was the best technique we were able to come up with. Of course, this sacrifices any 3rd-party registries and we were aware of that.

Could you please elaborate on what do you mean by the "SHA of the image"? Do you mean the SHA-256 hash at the end of the container image URLs? Do you mean that only the hash should be collected (without the URL part before the @ symbol)?

How should we filter Red Hat images if not based on the URL? You mention a "image reference code". This term doesn't seem to appear in the K8s API documentation, or at least not in the `client-go` code. Is that anoher name for the `ImageID` or is that something else entirely? Because we would certainly prefer to avoid collecting any data that we don't need and collecting thousands of customer images definitely falls under that category.

To conclude, as I said, the goal wasn't to collect everything, but rather a minimal sample that could later be expanded upon a further request. It may not have been a perfect solution, but I do believe it still provides some useful information without inflating the archives too much. We would have done the filtering differently, if only someone was able to tell us about a more sensible solution. This PR has been open for a very long time, making it available for a very wide audience to review. In the end I had trouble chasing down anyone willing to touch a PR into the I.O. repository, so I'm not sure what I was supposed to do to get a wider review. Surely enough, we would appreciate feedback from more people.

Reply:
> Could you please elaborate on what do you mean by the "SHA of the image"? Do you mean the SHA-256 hash at the end of the container image URLs? Do you mean that only the hash should be collected (without the URL part before the @ symbol)?

Yes, SHA is content-addressible (the same content always has that sha) and distinguishable via a database/dictionary lookup from the Red Hat side to identify images we have published without disclosing information about the contents of images we don't publish.

> How should we filter Red Hat images if not based on the URL? You mention a "image reference code". This term doesn't seem to appear in the K8s API documentation, or at least not in the client-go code. Is that anoher name for the ImageID or is that something else entirely?

`github.com/openshift/library-go/pkg/image/reference` is the canonical code within OpenShift to deal with image pull specs, you should never parse things but instead delegate to this logic.

> This PR has been open for a very long time, making it available for a very wide audience to review. In the end I had trouble chasing down anyone willing to touch a PR into the I.O. repository,

This is probably something I should have reviewed - I could have helped earlier had I known, but I will need someone to ping me in order to get attention.

I am currently drafting a doc in concert with other teams (engineering, business analytics, and will include CCX) to cover the larger subset of data we want like this (information about workloads).  I will share with you once that is available, and we can have a discussion about next steps that improve this process in the future.

Reply:
@smarterclayton any update on the doc draft to push this functionality forward?

Reply:
@smarterclayton Hi, it has been a while, but after some internal discussions, I have finally put together an attempt to improve the collection to better suit your requirements. Please see #174 and let me know what you think about it.

There seems to be a conflict between the following focuses:
- keeping the collected archive as small as possible
- including all Red Hat-related images in the gathered data
- scratching the URL-based filtering because it does not account for registry proxies

If we collect information about all running images, then the archive becomes way too large as the user's cluster grows and this is not good because, as far as I understand, we do not care about non-RH images at all, so information about them is meaningless. We were not able to find any reliable way of determining which images are produced/supported by Red Hat (we tried using labels, but that did not work out very well). The number of collected images must be limited in order to prevent archive from being too large, but this is makes it just as possible for Red Hat images to be missed as did using a URL-based filtering. If there are only few instances of each Red Hat image and many more of the end-user's own payloads, then we will not collect much of the information we care about.